File: SpinningHeart.java

import java.lang.Math;
import java.util.concurrent.TimeUnit;

public class SpinningHeart {
  public static void main(String[] main) throws InterruptedException {
    // Clear screen hide cursor
    System.out.print("\033[2J\033[?25l");

    float t = 0;
    while (true) {
      double zb[] = new double[100 * 40];
      double maxz = 0, c = Math.cos(t), s = Math.sin(t);
      for (double y = -0.5; y <= 0.5; y+= 0.01) {
        // Add beating effect
        double r = 0.4 + 0.05 * Math.pow(0.5 + 0.5 * Math.sin(t * 6 + y * 2), 8);
        for (double x = -0.5; x <= 0.5; x+= 0.01) {
          // heart formula
          double z = -x * x - Math.pow(1.2 * y - Math.abs(x) * 2 / 3, 2) + r * r;
          if (z < 0)
            continue;
          z = Math.sqrt(z) / (2 - y);
          for (double tz = -z; tz <= z; tz+= z / 6) {
            // Rotate
            double nx = x * c - tz * s;
            double nz = x * s + tz * c;

            // Add perspective
            double p = 1 + nz / 2;
            int vx = (int)Math.round((nx * p + 0.5) * 80 + 10);
            int vy = (int)Math.round((-y * p + 0.5) * 39 + 2);
            int idx = vx + vy * 100;
            if (zb[idx] <= nz) {
              zb[idx] = nz;
              if (maxz <= nz)
                maxz = nz;
            }
          }
        }
      }

      System.out.print("\033[H");
      for (int i = 0; i < 100 * 40; i++)
        System.out.print(i % 100 != 0 ? " .,-~:;=!*#$@@".charAt((int)Math.round(zb[i] / maxz * 13)) : 10);

      t+= 0.003f;
      TimeUnit.MILLISECONDS.sleep(3);
    }
  }
};

File: spinning_heart.py

import time, math

def main():
  # Clear screen hide cursor
  print("\x1b[2J\x1b[?25l", end='')
  zb = [0.0] * 100 * 40

  t = 0.0
  while True:
    maxz, c, s = 0, math.cos(t), math.sin(t)
    y = -0.5
    while y <= 0.5:
      # Add beating effect
      r = 0.4 + 0.05 * math.pow(0.5 + 0.5 * math.sin(t * 6 + y * 2), 8)
      x = -0.5
      while x <= 0.5:
        # heart formula
        z = -x * x - math.pow(1.2 * y - abs(x) * 2 / 3, 2) + r * r
        if z >= 0:
          z = math.sqrt(z) / (2 - y)
          tz = -z
          while tz <= z:
            # Rotate
            nx = x * c - tz * s
            nz = x * s + tz * c

            # Add perspective
            p = 1 + nz / 2
            vx = round((nx * p + 0.5) * 80 + 10)
            vy = round((-y * p + 0.5) * 39 + 2)
            idx = vx + vy * 100
            if zb[idx] <= nz:
              zb[idx] = nz
              if maxz < nz:
                maxz = nz
            tz+= z / 6
        x+= 0.01
      y+= 0.01

    print("\x1b[H", end='')
    for i in range(100 * 40):
      print(i % 100 and " .,-~:;=!*#$@@"[round(zb[i] / maxz * 13)] or "\n", end='')
      zb[i] = 0

    t+= 0.003
    time.sleep(0.003)

if __name__ == "__main__":
  main()
